homeassistant/components/emulated_kasa/__init__.py

```diff
@@ -0,0 +1,104 @@
+"""Support for local power state reporting of entities by emulating TP-Link Kasa smart plugs."""
+import logging
+from time import time
+
+from sense_energy import PlugInstance, SenseLink
+import voluptuous as vol
+
+import homeassistant.components as comps
+from homeassistant.const import (
+    ATTR_FRIENDLY_NAME,
+    CONF_ENTITIES,
+    CONF_NAME,
+    EVENT_HOMEASSISTANT_STOP,
+)
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.template import Template
+
+from .const import CONF_POWER, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_ENTITY_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Required(CONF_POWER): vol.Any(
+            vol.Coerce(float),
+            cv.template,
+        ),
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_ENTITIES): vol.Schema(
+                    {cv.entity_id: CONFIG_ENTITY_SCHEMA}
+                ),
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Sense component."""
+    hass.data.setdefault(DOMAIN, {})
+    conf = config.get(DOMAIN)
+    if not conf:
+        return True
+    hass.data[DOMAIN][CONF_ENTITIES] = conf.get(CONF_ENTITIES, {})
+
+    def devices():
+        """Drvices to be emulated."""
+        yield from get_plug_devices(hass)
+
+    server = SenseLink(devices)
+
+    async def stop_emulated_kasa(event):
+        await server.stop()
+
+    try:
+        await server.start()
+    except OSError as error:
+        _LOGGER.error("Failed to create UDP server at port 9999: %s", error)
+    else:
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_emulated_kasa)
+
+    return True
+
+
+def get_plug_devices(hass):
+    """Produce list of plug devices from config entities."""
+    entity_ids = hass.data[DOMAIN][CONF_ENTITIES]
+    for entity_id in entity_ids:
+        state = hass.states.get(entity_id)
+        if state is None:
```

Comment:
Should we log if the entity is missing?

Reply:
There's a chance it could be setup before the entity is available so that can add some extra misleading log entries.  We could add it as debug level

Reply:
We should wait for the STARTED event to setup like `HomeKit` does.  We will know the other integrations that are providing the state data will be setup.

Reply:
We can do that, however this part of the code will be called every time the devices are sent out so that would fill the logs up pretty quickly

Reply:
I added a check at startup that will print warnings if an entity isn't found

Reply:
It looks like you've taken care of that by validating once in start_emulated_kasa 👍